File: 09-decorator/09-decorator.py
from abc import ABC, abstractmethod

# Step 1: Define the Component Interface
class Shipping(ABC):
    @abstractmethod
    def cost(self) -> float:
        pass

    @abstractmethod
    def description(self) -> str:
        pass

# Step 2: Create a Concrete Component
class Standard(Shipping):
    def cost(self) -> float:
        return 5.0  # Standard shipping cost

    def description(self) -> str:
        return "Standard shipping"

# Step 3: Create an Abstract Decorator
class ShippingDecorator(Shipping):
    def __init__(self, shipping: Shipping):
        self.shipping = shipping

    def cost(self) -> float:
        return self.shipping.cost()

    def description(self) -> str:
        return self.shipping.description()

# Step 4: Create Concrete Decorators
class CashOnDelivery(ShippingDecorator):
    def cost(self) -> float:
        return super().cost() + 2.0  # Additional fee for COD

    def description(self) -> str:
        return super().description() + " with COD fee"

# Step 5: Use the Decorator
s = Standard()
cod = CashOnDelivery(s)

print(cod.description(), "costs $", cod.cost())

File: 09-decorator/09-decorator.java
interface Shipping {
    int cost();
    String description();
}

class Standard implements Shipping {
    @Override
    public int cost() {
        return 2;
    }

    @Override
    public String description() {
        return "Standard shipping";
    }
}

abstract class ShippingDecorator implements Shipping {
    protected Shipping shipping;

    public ShippingDecorator (Shipping shipping) {
        this.shipping = shipping;
    }

    @Override
    public int cost() {
        return this.shipping.cost();
    }

    @Override
    public String description() {
        return this.shipping.description();
    }
}

class ShippingWithFee extends ShippingDecorator {
    public ShippingWithFee(Shipping shipping) {
        super(shipping);
    }

    @Override
    public int cost() {
        return this.shipping.cost() + 1;
    }

    @Override
    public String description() {
        return this.shipping.description() + " with fee";
    }
}



public class Main {
    public static void main(String[] args) {
        Standard standard = new Standard();
        ShippingWithFee shippingWithFee = new ShippingWithFee(standard);
        System.out.println(shippingWithFee.description() + " " + shippingWithFee.cost());
    }
}

